node tests: Add SideEffect concept.

showell opened the pull request:
This avoids a kinda ugly idiom that is prevelant
in our codebase.

The key value provided here is that the developer
can specify **exactly** when they expect the
side effect to happen.

The code gives pretty nice error messages if
the side effect either happens too early or
too late (or never).

showell committed:
node tests: Add SideEffect concept.

This avoids a kinda ugly idiom that is prevelant
in our codebase.

The key value provided here is that the developer
can specify **exactly** when they expect the
side effect to happen.

The code gives pretty nice error messages if
the side effect either happens too early or
too late (or never).

showell committed:
node tests: Introduce CallbackData class.

This class avoids some action at a distance as well as
removing `let foo = bar` idioms, which are always a
code smell.

We can now more explicitly introduce const vars that
are based on callbacks closer to the code that consumes
them.

andersk commented:
Does this really add enough value to justify the cognitive overhead of a new bespoke abstraction that people interacting with the tests need to learn about?

Have you explored how other testing libraries approach this problem? (For example, Vitest [`expect(…).toHaveBeenCalled()`](https://vitest.dev/api/expect#tohavebeencalled), Sinon [expectations](https://sinonjs.org/releases/v19/mocks/#expectations), `node:test` [`MockFunctionContext.calls`](https://nodejs.org/api/test.html#ctxcalls).)

In `CallbackData`, `assert.ok(this.obj)` has confusing behavior if the data is a falsy primitive such as `undefined`, `null`, `false`, `0`, or `""`.

In `SideEffect`, it’s confusing to require the user to manually manage `reset()`. If there’s a `should_happen_during()` API, then the class should just track whether we’re inside `should_happen_during()` and prohibit occurrences of the effect outside (in which case `should_not_happen_during()` is redundant). Also, `has_happened()` sounds like the name of a function that returns a boolean.

showell commented:
The intention behind `CallbackData` is that `set()` and `get()` always deal with the JS equivalent of `Dict[str, object]`.  I can make the falsy check more specific, and we can eventually enforce types in TS.

showell commented:
I'll eliminate `reset`.  I agree that's kind of unnecessary.

showell committed:
list widget test: Avoid scroll_util leak with override.

I avoid scroll_util complications with my test_widget
helper.  I also no longer bother to verify that the
mock function gets called; for these particular tests,
it just isn't worth the trouble.

As often happens, the tests here are pointing out
code smells with the real code.  The list widget would
ideally have a nicer way to encapsulate all the strange
scroll logic.

showell commented:
I just pushed a new version.  I believe I have addressed all the concerns.

* I reduce cognitive overload by simplifying SideEffect.  There is no more reset.  CallbackData continues to just be set/get.
* I have explored some other testing frameworks and determined that `.should_be_called_during` is more robust than bracketing calls with `.not.toHaveBeenCalled` and `.toHaveBeenCalled`.  I also explained in https://chat.zulip.org/#narrow/channel/6-frontend/topic/node.20testing.20direction/near/2037781 why those options aren't even really easily available to us.
* I addressed the concerns about falsy checks in `CallbackData` with more rigorous checks.  (Also, in chat, I elaborated on my vision for how these will enable TS conversions.)
* I followed the wise suggestion from Anders to eliminate `reset` calls and be even more strict about preventing calls to `has_happened` outside of `should_happen_during` calls.  This led to me discovering a leak in the list_widget tests that are now being addressed on this PR.

Thanks for the feedback!  I think the new version is better.

showell committed:
activity test: Add scope blocks for a complicated test.

showell committed:
activity tests: Avoid possible leaks with fred.

We no longer impose the responsibility on individual
tests to restore fred's original data.

Note that people.add_active_user actually mutates
objects in place, hence the use of deep copies.
